fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryPredicateFilterPlan.java

```diff
@@ -0,0 +1,150 @@
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecord;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase;
+import com.apple.foundationdb.record.query.plan.temp.ExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.PlannerExpression;
+import com.apple.foundationdb.record.query.plan.temp.SingleExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.view.Source;
+import com.apple.foundationdb.record.query.plan.temp.view.SourceEntry;
+import com.apple.foundationdb.record.query.predicates.QueryPredicate;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A query plan that filters out records from a child plan that do not satisfy a {@link QueryPredicate}.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class RecordQueryPredicateFilterPlan extends RecordQueryFilterPlanBase {
+    @Nonnull
+    private final Source baseSource;
+    @Nonnull
+    private final ExpressionRef<QueryPredicate> filter;
+    @Nonnull
+    private final List<ExpressionRef<? extends PlannerExpression>> children;
+
+    public RecordQueryPredicateFilterPlan(@Nonnull RecordQueryPlan inner,
+                                          @Nonnull Source baseSource,
+                                          @Nonnull QueryPredicate filter) {
+        this(SingleExpressionRef.of(inner), baseSource, SingleExpressionRef.of(filter));
+    }
+
+    public RecordQueryPredicateFilterPlan(@Nonnull ExpressionRef<RecordQueryPlan> inner,
+                                          @Nonnull Source baseSource,
+                                          @Nonnull ExpressionRef<QueryPredicate> filter) {
+        super(inner);
+        this.baseSource = baseSource;
+        this.filter = filter;
+        this.children = ImmutableList.of(inner, filter);
+    }
+
+    @Override
+    protected boolean hasAsyncFilter() {
+        return false;
+    }
+
+    @Nullable
+    @Override
+    protected <M extends Message> Boolean evalFilter(@Nonnull FDBRecordStoreBase<M> store, @Nonnull EvaluationContext context, @Nullable FDBRecord<M> record) {
+        if (record == null) {
+            return null;
+        }
+        Boolean result = null;
+        Iterator<SourceEntry> entries = baseSource.evalSourceEntriesFor(record.getRecord()).iterator();
+        while (entries.hasNext()) {
+            Boolean entryResult = filter.get().eval(store, context, entries.next());
+            if (entryResult != null && entryResult) {
+                return true;
+            } else if (result == null) {
+                result = entryResult;
```

Comment:
We will produce more than one source entry when there's fan-out. I agree that the semantics here are wonky. Thankfully this will all be torn out and replaced in the next batch of changes with something that makes much more structural sense.